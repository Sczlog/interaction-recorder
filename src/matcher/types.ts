import { StateSchema } from 'xstate';
import {
  BlurEvent,
  ClickEvent,
  HoverEvent,
  KeydownEvent,
  KeypressEvent,
  KeyupEvent,
  MachineBeforeUnloadEvent,
  MachineWheelEvent,
  MousedownEvent,
  MousemoveEvent,
  MouseupEvent,
  ScrollEvent,
  TextChangeEvent,
  TextInputEvent,
} from '../types';
import { MatcherStep } from './index';

export interface MatcherContext {
  currentStep?: MatcherStep;
  previousStep?: MatcherStep;
}
/* eslint-disable @typescript-eslint/ban-types */
export interface MatcherSchema extends StateSchema {
  states: {
    INIT: {};
    CLICK: {};
    DRAG: {};
    TEXT: {};
    SCROLL: {};
    UNKNOWN: {};
  };
}
/* eslint-enable @typescript-eslint/ban-types */
export type MatcherEvent =
  | { type: 'mousedown'; data: MousedownEvent; target: HTMLElement | null }
  | { type: 'mouseup'; data: MouseupEvent; target: HTMLElement | null }
  | { type: 'click'; data: ClickEvent; target: HTMLElement | null }
  | { type: 'mousemove'; data: MousemoveEvent; target: HTMLElement | null }
  | { type: 'scroll'; data: ScrollEvent; target: HTMLElement | null }
  | { type: 'keydown'; data: KeydownEvent; target: HTMLElement | null }
  | { type: 'keypress'; data: KeypressEvent; target: HTMLElement | null }
  | { type: 'keyup'; data: KeyupEvent; target: HTMLElement | null }
  | { type: 'text_input'; data: TextInputEvent; target: HTMLElement | null }
  | { type: 'text_change'; data: TextChangeEvent; target: HTMLElement | null }
  | { type: 'blur'; data: BlurEvent; target: HTMLElement | null }
  | {
      type: 'before_unload';
      data: MachineBeforeUnloadEvent;
      target: HTMLElement | null;
    }
  | { type: 'hover'; data: HoverEvent; target: HTMLElement | null }
  | { type: 'wheel'; data: MachineWheelEvent; target: HTMLElement | null };

export type MatcherState =
  | {
      value: 'INIT';
      context: MatcherContext;
    }
  | {
      value: 'CLICK';
      context: MatcherContext;
    }
  | {
      value: 'DRAG';
      context: MatcherContext;
    }
  | {
      value: 'TEXT';
      context: MatcherContext;
    }
  | {
      value: 'SCROLL';
      context: MatcherContext;
    }
  | {
      value: 'UNKNOWN';
      context: MatcherContext;
    };
