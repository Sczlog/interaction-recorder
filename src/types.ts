import { IMeta } from './util/metaquerier';
import { IDataTransferItem } from './util/entry-reader';

export enum MatcherKey {
  RECEIVE_NEW_EVENT = 'matcher.browser_event.new',
  EMIT_NEW_STEP = 'matcher.step_event.new',
  EMIT_UPDATE_STEP = 'matcher.step_event.update',
  EMIT_END_STEP = 'matcher.step_event.end',
}

export type Step = {
  selector: IMeta;
  type:
    | 'CLICK'
    | 'RIGHT_CLICK'
    | 'DBLCLICK'
    | 'DRAG'
    | 'KEYPRESS'
    | 'TEXT'
    | 'BROWSE_FILE'
    | 'DROP_FILE'
    | 'NAVIGATION'
    | 'SCROLL'
    | 'REFRESH'
    | 'RESIZE'
    | 'UNKNOWN';
  events: StepEvent[];
};

export type StepEvent =
  | MouseEvents
  | ScrollEvent
  | KeydownEvent
  | KeypressEvent
  | TextInputEvent
  | TextChangeEvent
  | KeyupEvent
  | BlurEvent
  | MachineBeforeUnloadEvent
  | HoverEvent
  | MachineWheelEvent
  | DraggingEvent
  | DropEvent
  | DragStartEvent
  | DragEndEvent
  | DragEnterEvent
  | DragOverEvent
  | DragLeaveEvent
  | BrowseFileEvent;

export type Modifiers = {
  // only record modifers when needed
  ctrlKey?: true;
  altKey?: true;
  shiftKey?: true;
  metaKey?: true;
};

export type MousemoveRecord = {
  clientX: number;
  clientY: number;
  screenX: number;
  screenY: number;
  timeOffset: number;
};

type BaseEvent = {
  timestamp: number;
};
//#region mouse event
export type MachineMouseEvent = BaseEvent & {
  // we need to collect mouseEventInit;
  screenX: number;
  screenY: number;
  clientX: number;
  clientY: number;
  button: number;
  buttons: number;
  modifiers: Modifiers;
};

export type MousedownEvent = MachineMouseEvent & {
  type: 'mousedown';
};

export type MouseupEvent = MachineMouseEvent & {
  type: 'mouseup';
};

export type ClickEvent = MachineMouseEvent & {
  type: 'click';
};

export type DblClickEvent = MachineMouseEvent & {
  type: 'dblclick';
};

export type AuxClickEvent = MachineMouseEvent & {
  type: 'auxclick';
};

// mousemove is special
export type MousemoveEvent = BaseEvent & {
  type: 'mousemove';
  positions: Array<MousemoveRecord>;
};

type MouseEvents =
  | MousedownEvent
  | MouseupEvent
  | ClickEvent
  | DblClickEvent
  | AuxClickEvent
  | MousemoveEvent;
//#endregion mouse event

export type ScrollEvent = BaseEvent & {
  type: 'scroll';
  scrollLeft: number;
  scrollTop: number;
};

//#region keyboard event
export type MachineKeyboardEvent = BaseEvent & {
  key: string;
  code: string;
  keyCode: number;
  modifiers: Modifiers;
};

export type KeydownEvent = MachineKeyboardEvent & {
  type: 'keydown';
};

export type KeypressEvent = MachineKeyboardEvent & {
  type: 'keypress';
};

export type KeyupEvent = MachineKeyboardEvent & {
  type: 'keyup';
};
//#endregion keyboard event

//#region textinput
export type TextInputEvent = BaseEvent & {
  type: 'text_input';
  data: string;
  // departed
  value: string;
};

export type TextChangeEvent = BaseEvent & {
  type: 'text_change';
  value: string;
};
//#region textinput
export type BlurEvent = BaseEvent & {
  type: 'blur';
};

export type MachineBeforeUnloadEvent = BaseEvent & {
  type: 'before_unload';
};

export type HoverEvent = BaseEvent & {
  type: 'hover';
  clientX: number;
  clientY: number;
};

export type MachineWheelEvent = BaseEvent & {
  type: 'wheel';
};

export type BaseDragEvent = MachineMouseEvent & {
  targetIndex: number;
};

export type DraggingEvent = BaseDragEvent & {
  type: 'drag';
};
export type DragStartEvent = BaseDragEvent & {
  type: 'dragstart';
  effectAllowed: DataTransfer['effectAllowed'];
  items: Array<IDataTransferItem | undefined>;
};
export type DragEndEvent = BaseDragEvent & {
  type: 'dragend';
};
export type DragEnterEvent = BaseDragEvent & {
  type: 'dragenter';
};
export type DragOverEvent = BaseDragEvent & {
  type: 'dragover';
  dropEffect: DataTransfer['dropEffect'];
};
export type DragLeaveEvent = BaseDragEvent & {
  type: 'dragleave';
};
export type DropEvent = BaseDragEvent & {
  type: 'drop';
  effectAllowed: DataTransfer['effectAllowed'];
  dropEffect: DataTransfer['dropEffect'];
  items: Array<IDataTransferItem | undefined>;
};

export type BrowseFileEvent = BaseEvent & {
  type: 'file';
  files: Array<File>;
};
