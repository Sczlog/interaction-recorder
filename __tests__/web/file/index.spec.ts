import { Page } from 'puppeteer/lib/types';
import { IsolatedWorld } from '../isolatedWorld';
import path from 'path';
import 'jest-extended';

describe('file input', () => {
  let isolatedWorld: IsolatedWorld;
  beforeEach(async () => {
    isolatedWorld = new IsolatedWorld(page as unknown as Page);
    await page.goto(path.resolve(__dirname, 'file.html'), {
      waitUntil: 'domcontentloaded',
    });
    await isolatedWorld.injectRecorder();
    await isolatedWorld.createRecorder();
  });

  it('should handle file_input', async () => {
    await isolatedWorld.useFakeDateTime(100);
    const file_input = await page.$('input');
    // file_input?.click();
    await file_input?.uploadFile(path.resolve(__dirname, 'file.html'));
    const curr = await isolatedWorld.retrieveCurrentEvent();
    expect(curr.events).toMatchSnapshot();
    expect(curr.target).toMatchSnapshot({
      id: expect.any(String),
    });
    expect(curr.type).toEqualCaseInsensitive('browse_file');
  });

  // this version of cdp cannot drop file
  xit('should handle drop file', async () => {
    await isolatedWorld.useFakeDateTime(100);
    const file_input = await page.$('input');
    const bouding = await file_input?.boundingBox();
    isolatedWorld.client.send('Input.dispatchDragEvent', {
      type: 'drop',
      x: bouding!.x + 10,
      y: bouding!.y + bouding!.height / 2,
      data: {
        items: [],
        dragOperationsMask: 19,
      },
    });
    const curr = await isolatedWorld.retrieveCurrentEvent();
    expect(curr.events).toMatchSnapshot();
    expect(curr.target).toMatchSnapshot({
      id: expect.any(String),
    });
    expect(curr.type).toEqualCaseInsensitive('browse_file');
  });
});
