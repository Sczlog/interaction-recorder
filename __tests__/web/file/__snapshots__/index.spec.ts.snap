// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`file input should handle drop file 1`] = `
Array [
  Object {
    "files": Array [
      Object {},
    ],
    "timestamp": 100,
    "type": "file",
  },
]
`;

exports[`file input should handle drop file 2`] = `
Object {
  "attributes": Object {
    "type": "file",
  },
  "id": Any<String>,
  "tagName": "INPUT",
}
`;

exports[`file input should handle file_input 1`] = `
Array [
  Object {
    "files": Array [
      Object {},
    ],
    "timestamp": 100,
    "type": "file",
  },
]
`;

exports[`file input should handle file_input 2`] = `
Object {
  "attributes": Object {
    "type": "file",
  },
  "id": Any<String>,
  "tagName": "INPUT",
}
`;
